### PR TITLE
planner: disable cop mode for TiFlash if mpp mode is applicable

### LINKED ISSUE #37459: planner generates bad plan that uses cop[tiflash] instead of mpp[tiflash]
The plan of CHBenchmark-Q3 is still use coptask to query TiFlash. We should use mpp task for better performance.

```sql
mysql> explain analyze select   ol_o_id, ol_w_id, ol_d_id,          
sum(ol_amount) as revenue, o_entry_d 
from  customer, new_order, orders, order_line 
where  c_state like 'a%'   and c_id = o_c_id   and c_w_id = o_w_id   
and c_d_id = o_d_id   and no_w_id = o_w_id   and no_d_id = o_d_id   
and no_o_id = o_id   and ol_w_id = o_w_id   
and ol_d_id = o_d_id   and ol_o_id = o_id   
and o_entry_d > '2007-01-02 00:00:00.000000' 
group by ol_o_id, ol_w_id, ol_d_id, o_entry_d 
order by revenue desc, o_entry_d;
+----------------------------------------------+-------------+----------+--------------+-----------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------+
| id                                           | estRows     | actRows  | task         | access object                                                         | execution info                                                                                                                                                                                                                                                                                                                                                   | operator info                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | memory    | disk    |
+----------------------------------------------+-------------+----------+--------------+-----------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------+
| Sort_17                                      | 1.00        | 0        | root         |                                                                       | time:841.9ms, loops:1                                                                                                                                                                                                                                                                                                                                            | Column#47:desc, chbenchmark.orders.o_entry_d                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 0 Bytes   | 0 Bytes |
| └─Projection_19                              | 1.00        | 0        | root         |                                                                       | time:841.9ms, loops:1, Concurrency:OFF                                                                                                                                                                                                                                                                                                                           | chbenchmark.order_line.ol_o_id, chbenchmark.order_line.ol_w_id, chbenchmark.order_line.ol_d_id, Column#47, chbenchmark.orders.o_entry_d                                                                                                                                                                                                                                                                                                                                                                                   | 2.89 KB   | N/A     |
|   └─HashAgg_23                               | 1.00        | 0        | root         |                                                                       | time:841.9ms, loops:1, partial_worker:{wall_time:841.784952ms, concurrency:5, task_num:0, tot_wait:4.208651992s, tot_exec:0s, tot_time:4.208657376s, max:841.747087ms, p95:841.747087ms}, final_worker:{wall_time:841.841251ms, concurrency:5, task_num:0, tot_wait:4.208953897s, tot_exec:24.555µs, tot_time:4.208983488s, max:841.823867ms, p95:841.823867ms}  | group by:chbenchmark.order_line.ol_d_id, chbenchmark.order_line.ol_o_id, chbenchmark.order_line.ol_w_id, chbenchmark.orders.o_entry_d, funcs:sum(chbenchmark.order_line.ol_amount)->Column#47, funcs:firstrow(chbenchmark.orders.o_entry_d)->chbenchmark.orders.o_entry_d, funcs:firstrow(chbenchmark.order_line.ol_o_id)->chbenchmark.order_line.ol_o_id, funcs:firstrow(chbenchmark.order_line.ol_d_id)->chbenchmark.order_line.ol_d_id, funcs:firstrow(chbenchmark.order_line.ol_w_id)->chbenchmark.order_line.ol_w_id | 35.0 KB   | N/A     |
|     └─IndexJoin_34                           | 0.00        | 0        | root         |                                                                       | time:841.7ms, loops:1                                                                                                                                                                                                                                                                                                                                            | inner join, inner:IndexLookUp_33, outer key:chbenchmark.orders.o_w_id, chbenchmark.orders.o_d_id, chbenchmark.orders.o_id, inner key:chbenchmark.order_line.ol_w_id, chbenchmark.order_line.ol_d_id, chbenchmark.order_line.ol_o_id, equal cond:eq(chbenchmark.orders.o_d_id, chbenchmark.order_line.ol_d_id), eq(chbenchmark.orders.o_id, chbenchmark.order_line.ol_o_id), eq(chbenchmark.orders.o_w_id, chbenchmark.order_line.ol_w_id)                                                                                 | 0 Bytes   | N/A     |
|       ├─IndexJoin_100(Build)                 | 0.00        | 0        | root         |                                                                       | time:836.6ms, loops:1                                                                                                                                                                                                                                                                                                                                            | inner join, inner:IndexReader_99, outer key:chbenchmark.orders.o_w_id, chbenchmark.orders.o_d_id, chbenchmark.orders.o_id, inner key:chbenchmark.new_order.no_w_id, chbenchmark.new_order.no_d_id, chbenchmark.new_order.no_o_id, equal cond:eq(chbenchmark.orders.o_d_id, chbenchmark.new_order.no_d_id), eq(chbenchmark.orders.o_id, chbenchmark.new_order.no_o_id), eq(chbenchmark.orders.o_w_id, chbenchmark.new_order.no_w_id)                                                                                       | 0 Bytes   | N/A     |
|       │ ├─IndexJoin_118(Build)               | 0.00        | 0        | root         |                                                                       | time:836.5ms, loops:1                                                                                                                                                                                                                                                                                                                                            | inner join, inner:IndexLookUp_117, outer key:chbenchmark.customer.c_id, chbenchmark.customer.c_w_id, chbenchmark.customer.c_d_id, inner key:chbenchmark.orders.o_c_id, chbenchmark.orders.o_w_id, chbenchmark.orders.o_d_id, equal cond:eq(chbenchmark.customer.c_d_id, chbenchmark.orders.o_d_id), eq(chbenchmark.customer.c_id, chbenchmark.orders.o_c_id), eq(chbenchmark.customer.c_w_id, chbenchmark.orders.o_w_id)                                                                                                  | 0 Bytes   | N/A     |
|       │ │ ├─TableReader_163(Build)           | 0.00        | 0        | root         |                                                                       | time:836.3ms, loops:1, cop_task: {num: 311, max: 206ms, min: 18.3ms, avg: 40ms, p95: 72.7ms, rpc_num: 311, rpc_time: 12.4s, copr_cache_hit_ratio: 0.00, distsql_concurrency: 15}                                                                                                                                                                                 | data:Selection_162                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | 926 Bytes | N/A     |
|       │ │ │ └─Selection_162                  | 0.00        | 0        | cop[tiflash] |                                                                       | tiflash_task:{proc max:203.7ms, min:16.7ms, avg: 38.6ms, p80:38.4ms, p95:71.5ms, iters:0, tasks:311, threads:311}                                                                                                                                                                                                                                                | like(chbenchmark.customer.c_state, "a%", 92)                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | N/A       | N/A     |
|       │ │ │   └─TableFullScan_161            | 45000000.00 | 45000000 | cop[tiflash] | table:customer                                                        | tiflash_task:{proc max:199.7ms, min:12.7ms, avg: 34.4ms, p80:33.9ms, p95:68.5ms, iters:944, tasks:311, threads:311}                                                                                                                                                                                                                                              | keep order:false                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | N/A       | N/A     |
|       │ │ └─IndexLookUp_117(Probe)           | 0.00        | 0        | root         |                                                                       |                                                                                                                                                                                                                                                                                                                                                                  |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | N/A       | N/A     |
|       │ │   ├─Selection_115(Build)           | 0.00        | 0        | cop[tikv]    |                                                                       |                                                                                                                                                                                                                                                                                                                                                                  | not(isnull(chbenchmark.orders.o_c_id))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | N/A       | N/A     |
|       │ │   │ └─IndexRangeScan_113           | 0.00        | 0        | cop[tikv]    | table:orders, index:idx_order(o_w_id, o_d_id, o_c_id, o_id)           |                                                                                                                                                                                                                                                                                                                                                                  | range: decided by [eq(chbenchmark.orders.o_w_id, chbenchmark.customer.c_w_id) eq(chbenchmark.orders.o_d_id, chbenchmark.customer.c_d_id) eq(chbenchmark.orders.o_c_id, chbenchmark.customer.c_id)], keep order:false                                                                                                                                                                                                                                                                                                      | N/A       | N/A     |
|       │ │   └─Selection_116(Probe)           | 0.00        | 0        | cop[tikv]    |                                                                       |                                                                                                                                                                                                                                                                                                                                                                  | gt(chbenchmark.orders.o_entry_d, 2007-01-02 00:00:00.000000)                                                                                                                                                                                                                                                                                                                                                                                                                                                              | N/A       | N/A     |
|       │ │     └─TableRowIDScan_114           | 0.00        | 0        | cop[tikv]    | table:orders                                                          |                                                                                                                                                                                                                                                                                                                                                                  | keep order:false                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | N/A       | N/A     |
|       │ └─IndexReader_99(Probe)              | 0.00        | 0        | root         |                                                                       |                                                                                                                                                                                                                                                                                                                                                                  | index:IndexRangeScan_98                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | N/A       | N/A     |
|       │   └─IndexRangeScan_98                | 0.00        | 0        | cop[tikv]    | table:new_order, index:PRIMARY(no_w_id, no_d_id, no_o_id)             |                                                                                                                                                                                                                                                                                                                                                                  | range: decided by [eq(chbenchmark.new_order.no_w_id, chbenchmark.orders.o_w_id) eq(chbenchmark.new_order.no_d_id, chbenchmark.orders.o_d_id) eq(chbenchmark.new_order.no_o_id, chbenchmark.orders.o_id)], keep order:false                                                                                                                                                                                                                                                                                                | N/A       | N/A     |
|       └─IndexLookUp_33(Probe)                | 0.00        | 0        | root         |                                                                       |                                                                                                                                                                                                                                                                                                                                                                  |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | N/A       | N/A     |
|         ├─IndexRangeScan_31(Build)           | 0.00        | 0        | cop[tikv]    | table:order_line, index:PRIMARY(ol_w_id, ol_d_id, ol_o_id, ol_number) |                                                                                                                                                                                                                                                                                                                                                                  | range: decided by [eq(chbenchmark.order_line.ol_w_id, chbenchmark.orders.o_w_id) eq(chbenchmark.order_line.ol_d_id, chbenchmark.orders.o_d_id) eq(chbenchmark.order_line.ol_o_id, chbenchmark.orders.o_id)], keep order:false                                                                                                                                                                                                                                                                                             | N/A       | N/A     |
|         └─TableRowIDScan_32(Probe)           | 0.00        | 0        | cop[tikv]    | table:order_line                                                      |                                                                                                                                                                                                                                                                                                                                                                  | keep order:false                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | N/A       | N/A     |
+----------------------------------------------+-------------+----------+--------------+-----------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+---------+
19 rows in set (0.91 sec)
```

```sql
mysql> select * from information_schema.cluster_info;
+---------+------------------+-------------------+-------------+------------------------------------------+---------------------------+------------------+-----------+
| TYPE    | INSTANCE         | STATUS_ADDRESS    | VERSION     | GIT_HASH                                 | START_TIME                | UPTIME           | SERVER_ID |
+---------+------------------+-------------------+-------------+------------------------------------------+---------------------------+------------------+-----------+
| tidb    | 172.16.5.82:8224 | 172.16.5.82:8724  | 6.3.0-alpha | 0d5c56699f514cb5b65f7dbd2cf5e2b60ae1bf01 | 2022-08-29T16:19:04+08:00 | 4m17.470205756s  |    670693 |
| pd      | 172.16.5.82:6724 | 172.16.5.82:6724  | 6.3.0-alpha | 2adb9797473844e094a4571e7ca2e5388187681b | 2022-08-29T15:51:59+08:00 | 31m22.470215233s |         0 |
| tikv    | 172.16.5.82:7724 | 172.16.5.82:16724 | 6.2.0-alpha | 0030aeb90a840140a935fbb0181b6a62b5e680b3 | 2022-08-29T15:52:37+08:00 | 30m44.470218023s |         0 |
| tiflash | 172.16.5.85:5824 | 172.16.5.85:17224 | 6.3.0-alpha | b109a12ec8ae99698110d44ad747d7b3a9bdd9fd | 2022-08-29T15:52:41+08:00 | 30m40.470220217s |         0 |
+---------+------------------+-------------------+-------------+------------------------------------------+---------------------------+------------------+-----------+
4 rows in set (0.03 sec)
```

_Originally posted by @JinheLin in https://github.com/pingcap/tidb/issues/37422#issuecomment-1229949169_

### PR DESCRIPTION
**Pull request by fixdb**
### What problem does this PR solve?
Disable cop mode for TiFlash if MPP is applicable

Issue Number: close #37459

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -1957,6 +1971,10 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 		mppTask = ts.addPushedDownSelectionToMppTask(mppTask, ds.stats.ScaleByExpectCnt(prop.ExpectedCnt))
 		return mppTask, nil
 	}
+	if ts.StoreType == kv.TiFlash && mppApplicable && ds.SCtx().GetSessionVars().IsMPPAllowed() {
+		// don't generate cop task for TiFlash if mpp task is applicable
+		return invalidTask, nil
```
Comment (AilinKid) [inline planner/core/find_best_task.go:1976]: this is why? two tasks can't co-exist together? or the cop task can always win the cost comparison

Comment (fixdb) [reply planner/core/find_best_task.go:1976]: Can two tasks can co-exist. But we always prefer MPP tasks if it is applicable. But our cost model may not be able to make the correct decision and choose cop task as the final plan, which is not what we want.

**Comment by AilinKid** [comment]
https://github.com/pingcap/tidb/pull/44249 since convert2TableScan has changed a lot, the code is refined here

**Comment by fixdb** [comment]
Closing this PR in favor of https://github.com/pingcap/tidb/pull/44249

**Pull request closed without merging**